framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java

```diff
@@ -0,0 +1,372 @@
+package org.checkerframework.framework.type;
+
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.framework.qual.PolyAll;
+import org.checkerframework.framework.util.AnnotatedTypes;
+import org.checkerframework.framework.util.AnnotationBuilder;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+import java.net.JarURLConnection;
+import java.net.URL;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.AnnotationMirror;
+
+/*>>>
+import org.checkerframework.checker.nullness.qual.Nullable;
+ */
+
+/**
+ * This class assists the {@link AnnotatedTypeFactory} by reflectively looking
+ * up the list of qualifier class names in each checker's qual directory, and
+ * then loading and returning it as a set of Annotation Class instances. It can
+ * also look up and load qualifiers from external directories that are passed as
+ * arguments to checkers that have extension capabilities such as the Subtyping
+ * Checker or Units Checker.
+ *
+ * Only annotation classes that have the {@link Target} meta-annotation with the
+ * values of {@link ElementType#TYPE_USE} and/or
+ * {@link ElementType#TYPE_PARAMETER} are loaded. Other type qualifiers must be
+ * manually listed in a checker's annotated type factory by overriding
+ * {@link AnnotatedTypeFactory#createSupportedTypeQualifiers()}.
+ * 
+ *
+ * The methods in this class are called within
+ * {@link AnnotatedTypeFactory#loadTypeQualifiersFromQualDir(boolean, Set)}.
+ *
+ * Type system checker writers may wish to subclass this class if they wish to
+ * implement some custom rules to filter or process loaded annotation classes,
+ * by providing an override implementation of
+ * {@link #createAnnotationMirrorFromClass(Class)}
+ * 
+ * @author jeff luo
+ */
+public class AnnotatedTypeLoader {
+    private BaseTypeChecker checker;
+
+    // For loading from a source package directory
+    private String packageName;
+    private static final String QUAL_PACKAGE_SUFFIX = ".qual";
+
+    // For loading from a Jar file
+    private static final String CLASS_SUFFIX = ".class";
+
+    // For loading from external directories
+    private static final String JAVA_SUFFIX = ".java";
+
+    // Constants
+    private static final char DOT = '.';
+    private static final char SLASH = '/';
```

Comment:
No, because it's in a jar file.  We should add a global  variable for this so we don't keep forget this.